docs/reference/developer/adr/0017-restrict-dataset-to-items-with-matching-dimensionality.rst

```diff
@@ -0,0 +1,97 @@
+- Complicated logic for selecting coordinates, affecting code as well as documentation needs.
+  If a dataset has a dimension of length 1, should items without that dim have the coordinate?
+- Slicing a dim can make coords suddenly apply for an item.
+- Readonly flags in ``DataArray`` are required to avoid modifying data in lower-dimensional items via slices of a dataset.
+- Consistency issues after the introduction of the alignment flag: `issues selecting coords with length-2 bin edges #3148 <https://github.com/scipp/scipp/issues/3148>`_ and `disappearing unaligned coords #3149 <https://github.com/scipp/scipp/issues/3149>`_.
+- ``Dataset.dim`` may "lie" if an item is 0-D.
+  Maybe it should have raised unless all items are 1-D?
+- ``dims`` and ``sizes`` of ``DataArray`` imply an order, but they do not for ``Dataset``.
+  This leads to some code overhead and risk of confusion.
+- Complicated (internal) logic for updating the ``sizes`` dict.
+  This is not a problem for the user, except for rare edge cases where size-changing item replacements are not support although they could be.
+
+Given the long-term focus of the project, and the limited area of applicability of ``Dataset`` with its current semantics, we believe that it is worth considering a change.
+While none of the above issues is major, they add up to a significant amount of complexity that may turn out hard to manage or justify in the long run.
+Even in its current state several aspects of the above are not well-documented, neither for developers nor for users.
+
+Proposed solution
+~~~~~~~~~~~~~~~~~
+
+We propose to restrict ``Dataset`` to items with matching dimensionality.
+Each item of a dataset will retain a ``masks`` dictionary (no change).
+
+We can then think of Scipp providing a natural cascade of objects:
+Given a number of arrays that we would like to combine, if ``sizes`` and ``dtype`` are consistent use ``DataArray`` with an extra dimension.
+If only ``sizes`` are consistent use ``Dataset``.
+If ``sizes`` are not consistent use ``DataGroup``.
+
+Decision
+--------
+
+- ``Dataset`` will be restricted to items with matching dimensionality.
+
+Consequences
+------------
+
+Positive:
+~~~~~~~~~
+
+- Cleaner and simpler semantics.
+- Simplified documentation and training.
+- Simplified code.
+- Resolves a number of issues where semantics of ``Dataset`` are not clear in edge-cases, in particular with bin-edge coordinates.
+
+Negative:
+~~~~~~~~~
+
+- ``Dataset`` will no longer be able to represent certain types of data.
+  Users will need to resort to ``DataGroup`` instead, which has other limitations, such as requiring to duplicate coordinates.
+  Another option would be to replicate data values of the lower-dimensional items to match the dimensionality of the higher-dimensional items.
+  This would reuqire more memory, but would force the users to be explicit about the meaning of data they want to represent.
```

Comment:
This could be done with a broadcast which would only increase memory use when you do an operation with it.